consensus/src/simplex/actors/voter/actor.rs

```diff
@@ -1221,311 +594,360 @@ impl<
-        &mut self,
-        view: u64,
-        force: bool,
-    ) -> Option<Notarization<S, D>> {
-        // Get requested view
-        let round = self.views.get_mut(&view)?;
+    fn construct_finalize(&mut self, view: u64) -> Option<Finalize<S, D>> {
+        // Determine if it makes sense to broadcast a finalize
+        let proposal = self.state.finalize_candidate(view)?;
 
-        // Attempt to construct notarization
-        round.notarizable(force).await
+        // Construct finalize
+        Finalize::sign(self.state.scheme(), &self.namespace, proposal)
     }
 
-    async fn construct_nullification(
-        &mut self,
-        view: u64,
-        force: bool,
-    ) -> Option<Nullification<S>> {
-        // Get requested view
-        let round = self.views.get_mut(&view)?;
+    fn construct_notarization(&mut self, view: u64, force: bool) -> Option<Notarization<S, D>> {
+        let mut timer = self.recover_latency.timer();
+        match self.state.notarization_candidate(view, force) {
+            Some((new, notarization)) => {
+                if new {
+                    timer.observe();
+                } else {
+                    timer.cancel();
+                }
+                Some(notarization)
+            }
+            None => {
+                timer.cancel();
+                None
+            }
+        }
+    }
 
-        // Attempt to construct nullification
-        round.nullifiable(force).await
+    fn construct_nullification(&mut self, view: u64, force: bool) -> Option<Nullification<S>> {
+        let mut timer = self.recover_latency.timer();
+        match self.state.nullification_candidate(view, force) {
+            Some((new, nullification)) => {
+                if new {
+                    timer.observe();
+                } else {
+                    timer.cancel();
+                }
+                Some(nullification)
+            }
+            None => {
+                timer.cancel();
+                None
+            }
+        }
     }
 
-    fn construct_finalize(&mut self, view: u64) -> Option<Finalize<S, D>> {
-        // Determine if it makes sense to broadcast a finalize
-        let round = self.views.get_mut(&view)?;
-        if round.broadcast_nullify {
-            return None;
+    fn construct_finalization(&mut self, view: u64, force: bool) -> Option<Finalization<S, D>> {
+        let mut timer = self.recover_latency.timer();
+        match self.state.finalization_candidate(view, force) {
+            Some((new, finalization)) => {
+                if new {
+                    timer.observe();
+                } else {
+                    timer.cancel();
+                }
+                Some(finalization)
+            }
+            None => {
+                timer.cancel();
+                None
+            }
         }
-        if !round.broadcast_notarize {
-            // Ensure we notarize before we finalize
-            return None;
+    }
```

Comment:
minor suggestion here and in the other `construct_*` methods

```suggestion
    fn construct_finalization(&mut self, view: u64, force: bool) -> Option<Finalization<S, D>> {
        let mut timer = self.recover_latency.timer();
        let candidate = self.state.finalization_candidate(view, force) {
        if let Some((true, _)) = candidate {
            timer.observe();
        } else {
            timer.cancel();
        }
        candidate.map(|(_, finalization)| finalization)
    }
```